src/drivers/imx/interrupt.c

```diff
@@ -3,26 +3,413 @@
+}
+
+static inline void handle_irq_batch(struct irq_cascade_desc *cascade,
+				    uint32_t line_index, uint64_t status)
+{
+	int core = cpu_get_id();
+	struct list_item *clist;
+	struct irq_desc *child = NULL;
+
+	while (status) {
+		int bit = get_first_irq(status);
+		bool handled = false;
+
+		status &= ~(1 << bit); /* Release interrupt */
+
+		spin_lock(cascade->lock);
+
+		/* Get child if any and run handler */
+		list_for_item(clist, &cascade->child[bit].list) {
+			child = container_of(clist, struct irq_desc, irq_list);
+
+			if (child->handler && (child->cpu_mask & 1 << core)) {
+				child->handler(child->handler_arg);
+				handled = true;
+			}
+		}
+
+		spin_unlock(cascade->lock);
+
+		if (!handled) {
+			trace_irq_error("irq_handler(): nobody cared, bit %d",
+					bit);
+			/* Mask this interrupt so it won't happen again */
+			irqstr_mask_int(line_index * 64 + bit);
+		}
+	}
+}
+
+static inline void irq_handler(void *data, uint32_t line_index)
+{
+	struct irq_desc *parent = data;
+	struct irq_cascade_desc *cascade =
+		container_of(parent, struct irq_cascade_desc, desc);
+	uint64_t status;
+	uint32_t tries = IRQ_MAX_TRIES;
+
+	status = get_irqsteer_interrupts(line_index);
+
+	while (status) {
+		/* Handle current interrupts */
+		handle_irq_batch(cascade, line_index, status);
+
+		/* Any interrupts happened while we were handling the
+		 * current ones?
+		 */
+		status = get_irqsteer_interrupts(line_index);
+		if (!status)
+			break;
+
+		/* Any device keeping interrupting while we're handling
+		 * or can't clear?
+		 */
+
+		if (!--tries) {
+			tries = IRQ_MAX_TRIES;
+			trace_irq_error("irq_handler(): IRQ storm, status "
+					PRIx64,
+					get_irqsteer_interrupts(line_index));
+		}
+	}
+}
+
+#define DEFINE_IRQ_HANDLER(n) \
+	static inline void irqstr_irqhandler_##n(void *arg) \
+	{ \
+		irq_handler(arg, n); \
+	}
+
+DEFINE_IRQ_HANDLER(0)
+DEFINE_IRQ_HANDLER(1)
+DEFINE_IRQ_HANDLER(2)
+DEFINE_IRQ_HANDLER(3)
+DEFINE_IRQ_HANDLER(4)
+DEFINE_IRQ_HANDLER(5)
+DEFINE_IRQ_HANDLER(6)
+DEFINE_IRQ_HANDLER(7)
+
+static void irq_mask(struct irq_desc *desc, uint32_t irq, unsigned int core)
+{
+	uint32_t irq_base = desc->irq - IRQ_NUM_IRQSTR_DSP0;
+
+	(void)core;
+
+	irq_base *= 64;
+	irq += irq_base;
+
+	/* This is the actual IRQ_STEER IRQ number */
+
+	irqstr_mask_int(irq);
+}
+
+static void irq_unmask(struct irq_desc *desc, uint32_t irq, unsigned int core)
+{
+	uint32_t irq_base = desc->irq - IRQ_NUM_IRQSTR_DSP0;
+
+	(void)core;
+
+	irq_base *= 64;
+	irq += irq_base;
+
+	/* This is the actual IRQ_STEER IRQ number */
+
+	irqstr_unmask_int(irq);
+}
+
+static const struct irq_cascade_ops irq_ops = {
+	.mask = irq_mask,
+	.unmask = irq_unmask,
+};
+
+/* IRQ_STEER interrupts */
+#define IRQSTR_CASCADE_TMPL_DECL(n) \
+	{ \
+		.name = irq_name_irqsteer[n], \
+		.irq = IRQ_NUM_IRQSTR_DSP##n, \
+		.handler = irqstr_irqhandler_##n, \
+		.ops = &irq_ops, \
+		.global_mask = false, \
+	},
+static const struct irq_cascade_tmpl dsp_irq[] = {
```

Comment:
Add empty line.

Reply:
Oops, not sure why I didn't have it. Pushing new version, addressing most of these comments.